### PR TITLE
[V2] tpm2_createpolicy: support to save session context with the option -S

### PR DESCRIPTION
**Pull request by jiazhang0**
The new option -S, associated with either -e or -a option, allows to
save the session context to a file.

Signed-off-by: Jia Zhang <qianyue.zj@alibaba-inc.com>

### CODE REVIEW

File: tools/tpm2_createpolicy.c [hunk 1]
```diff
@@ -180,6 +176,12 @@ static bool on_option(char key, char *value) {
     case 'e':
         pctx.common_policy_options.extend_policy_session = true;
         break;
+    case 'C':
+        pctx.common_policy_options.context_file = value;
+        if (value == NULL || value[0] == '\0') {
+            return false;
+        }
```
Comment (williamcroberts) [inline tools/tpm2_createpolicy.c]: value == NULL impossible and an empty string, who cares, no need for this check.

File: tools/tpm2_createpolicy.c [hunk 2]
```diff
@@ -213,5 +216,23 @@ int tpm2_tool_onrun(TSS2_SYS_CONTEXT *sapi_context, tpm2_option_flags flags) {
...
+    if (parse_policy_type_specific_command(sapi_context) != TPM2_RC_SUCCESS) {
+        return 1;
+    }
+
+    if (pctx.common_policy_options.extend_policy_session) {
+        TPM2_HANDLE handle;
+
+        handle = pctx.common_policy_options.policy_session->sessionHandle;
+
+        tpm2_tool_output("EXTENDED_POLICY_SESSION_HANDLE: 0x%08X\n", handle);
```
Comment (williamcroberts) [inline tools/tpm2_createpolicy.c]: all output to stdout should be done in YAML now.

File: man/tpm2_createpolicy.1.md [hunk 3]
```diff
@@ -46,6 +46,9 @@ These options control creating the policy authorization session:
     Start a policy session of type **TPM_SE_POLICY**. Default without this option
     is **TPM_SE_TRIAL**.
 
+  * **-C**, **--context**:_CONTEXT\_FILE_:
+    An optional file used to store the session context created by either -e or -a.
+
```
Comment (williamcroberts) [inline man/tpm2_createpolicy.1.md]: Not sure on this. I was hoping that abrmd is getting session continuation support you would just chain everything together with -S <FILE|HANDLE>.

```
tpm2_createpolicy -S foo ....
tpm2_<command> -S foo ...
...
```
On some commands, like tpm2_load,  -C is used for object context handle save. I think I want this to be -S, so we can start moving the tools to support more and more session handling.

**Comment by jiazhang0** [comment]
@williamcroberts 

A quick diff against V1:

```
diff --git a/man/tpm2_createpolicy.1.md b/man/tpm2_createpolicy.1.md
index e56328f..de16454 100644
--- a/man/tpm2_createpolicy.1.md
+++ b/man/tpm2_createpolicy.1.md
@@ -46,8 +46,9 @@ These options control creating the policy authorization session
     Start a policy session of type **TPM_SE_POLICY**. Default without this optio
     is **TPM_SE_TRIAL**.
 
-  * **-C**, **--context**:_CONTEXT\_FILE_:
-    An optional file used to store the session context created by either -e or -
+  * **-S**, **--save-session-context**:_CONTEXT\_FILE_:
+    An optional file used to store the session context created by either -e
+    or -a.
 
 [common options](common/options.md)
 
diff --git a/tools/tpm2_createpolicy.c b/tools/tpm2_createpolicy.c
index 573c6b9..d243073 100644
--- a/tools/tpm2_createpolicy.c
+++ b/tools/tpm2_createpolicy.c
@@ -178,9 +178,6 @@ static bool on_option(char key, char *value) {
         break;
     case 'C':
         pctx.common_policy_options.context_file = value;
-        if (value == NULL || value[0] == '\0') {
-            return false;
-        }
         break;
     }
 
@@ -197,10 +194,10 @@ bool tpm2_tool_onstart(tpm2_options **opts) {
         { "policy-pcr",     no_argument,        NULL,   'P' },
         { "auth-policy-session", no_argument, NULL,     'a'},
         { "extend-policy-session", no_argument, NULL,   'e'},
-        { "context",               required_argument, NULL, 'C' },   
+        { "save-session-context",  required_argument, NULL, 'S' },   
     };
 
-    *opts = tpm2_options_new("f:g:L:F:PaeC:", ARRAY_LEN(topts), topts, on_option
+    *opts = tpm2_options_new("f:g:L:F:PaeS:", ARRAY_LEN(topts), topts, on_option
 
     return *opts != NULL;
 }
@@ -225,12 +222,13 @@ int tpm2_tool_onrun(TSS2_SYS_CONTEXT *sapi_context, tpm2_op
 
         handle = pctx.common_policy_options.policy_session->sessionHandle;
 
-        tpm2_tool_output("EXTENDED_POLICY_SESSION_HANDLE: 0x%08X\n", handle);
+        LOG_INFO("EXTENDED_POLICY_SESSION_HANDLE: 0x%08X\n", handle);
 
         const char *file = pctx.common_policy_options.context_file;
 
         if (file) {
-            return files_save_tpm_context_to_file(sapi_context, handle, file) !=
+            return files_save_tpm_context_to_file(sapi_context,
+                                                  handle, file) != true;
         }
     }
```

The intention of this patch is to allow the incoming tpm2_flushcontext to flush a loaded or saved session context. So I didn't consider the plan of using -S to chain the saved session context file.

**Comment by williamcroberts** [comment]
LGTM please rebase and handle conflicts.

**Comment by williamcroberts** [comment]
moved:
https://github.com/intel/tpm2-tools/pull/667

**Pull request closed without merging**